File: array/a8.py
# Topic:  find Largest sum contiguous Subarray 
from typing import List, Any


class Array:

    def __init__(self, elements: List[Any]) -> None:
        self.elements = elements

    def sub_max(self):
        _max = self.elements[0]
        max_now = self.elements[0]

        for element in self.elements[1:]:
            max_now = max(element, max_now + element)
            _max = max(max_now, _max)
        return _max


array = Array([1, 2, 3, -2, 5])
print(array.sub_max())


def test_a8():
    array = Array([-2, 1, -3, 4, -1, 2, 1, -5, 4])
    assert 6 == array.sub_max()

    array = Array([1, 2, 3, -2, 5])
    assert 9 == array.sub_max()

    array = Array([-1, -2, -3, -4])
    assert -1 == array.sub_max()

File: array/a8.rs
// Topic: find Largest sum contiguous Subarray 

#[derive(PartialEq, PartialOrd)]
struct Array { elements: Vec<i64> }

impl Array {

    fn max_value(&self, first: i64, second: i64) -> i64 {
        if first > second {
            return first;
        } else if second > first {
            return second;
        } else {
            return first;
        }
    }

    fn sub_max(&self) -> i64 {
        let mut max = self.elements[0];
        let mut max_till_now = self.elements[0];

        for element in self.elements[1..].iter() {
            let element = *element;
            max_till_now = self.max_value(element, max_till_now + element);
            max = self.max_value(max_till_now, max);
        }
        return max;
    }
}

fn main() {
    let array = Array { elements: vec![-2, 1, -3, 4, -1, 2, 1, -5, 4]  };
    println!("{:?}", array.sub_max());
}
